test/e2e_node/node_shutdown_linux_test.go

```diff
@@ -0,0 +1,202 @@
+				getGracePeriodOverrideTestPod("period-5", nodeName, 5, false),
+				getGracePeriodOverrideTestPod("period-critical-120", nodeName, 120, true),
+				getGracePeriodOverrideTestPod("period-critical-5", nodeName, 5, true),
+			}
+
+			// Create pods
+			f.PodClient().CreateBatch(pods)
+
+			list, err := f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				framework.ExpectEqual(
+					pod.Status.Phase == v1.PodRunning,
+					true,
+					"pod is not ready",
+				)
+			}
+
+			// Emit PrepareForShutdown signal
+			err = emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+
+			time.Sleep(5 * time.Second)
+			// Not critical pod should be shutdown
+			list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				if kubelettypes.IsCriticalPod(&pod) {
+					framework.ExpectEqual(
+						pod.Status.Phase == v1.PodRunning,
+						true,
+						"critical pod should not be shutdown",
+					)
+				} else {
+					framework.ExpectEqual(
+						pod.Status.Phase == v1.PodFailed && pod.Status.Reason == "Shutdown",
+						true,
+						"pod should be shutdown",
+					)
+				}
+			}
+
+			time.Sleep(6 * time.Second)
+			// All pod should be shutdown
+			list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				framework.ExpectEqual(
+					pod.Status.Phase == v1.PodFailed && pod.Status.Reason == "Shutdown",
+					true,
+					"pod should be shutdown",
+				)
+			}
+		})
+
+		ginkgo.It("Interrupt shutdown", func() {
+			err := emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+			for i := 0; ; i++ {
+				isReady := getNodeReadyStatus(f)
+				if !isReady {
+					break
+				}
+				if i == 2 {
+					framework.ExpectEqual(
+						isReady,
+						false,
+						"node did not become shutdown as expected",
+					)
+				} else {
+					time.Sleep(time.Second)
+				}
+			}
+
+			err = emitSignalPrepareForShutdown(false)
+			framework.ExpectNoError(err)
+			for i := 0; ; i++ {
+				isReady := getNodeReadyStatus(f)
+				if isReady {
+					break
+				}
+				if i == 10 {
+					framework.ExpectEqual(
+						isReady,
+						true,
+						"node did not recover as expected",
+					)
+				} else {
+					time.Sleep(time.Second)
+				}
+			}
+		})
+	})
+})
+
+func getGracePeriodOverrideTestPod(name string, node string, gracePeriod int64, critical bool) *v1.Pod {
+	pod := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    name,
+					Image:   busyboxImage,
+					Command: []string{"sh", "-c"},
+					Args: []string{`
+_term() {
+	echo "Caught SIGTERM signal!"
+	sleep 300
+}
+trap _term SIGTERM 
+sleep 300
```

Comment:
pod will exist if test somehow gets stuck and takes a while (> than 300s). maybe bump up from 300s to say 10mins or something like that? not sure what's reasonable time for test to execute :)